### PR TITLE
Add RSEM prepare reference component

### PR DESCRIPTION
**Pull request by sainirmayi**
## Issue ticket number

Closes #xxxx 

## Checklist before requesting a review

- [ ] I have performed a self-review of my code

- [ ] Conforms to the [Contributing guidelines](https://github.com/viash-hub/base/blob/main/CONTRIBUTING.md)

- [ ] Proposed changes are described in the [CHANGELOG.md](https://github.com/viash-hub/base/blob/main/CHANGELOG.md)

- [ ] I have tested my code with `viash ns test --parallel -q <name or namespace>`

- Check the correct box. Does this PR contain:
  - [ ] Breaking changes
  - [x] New functionality
  - [ ] Major changes
  - [ ] Minor changes
  - [ ] Documentation
  - [ ] Bug fixes

### CODE REVIEW

**Comment by rcannood** [comment]
@sainirmayi Could you merge main into your branch? ^^

File: src/rsem/rsem_prepare_reference/config.vsh.yaml [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+  - name: Other options
+    arguments: 
+      - name: --gtf
+        type: file
+        description: Assume that 'reference_fasta_files' contains the sequence of a genome, and extract transcript reference sequences using the gene annotations specified in the GTF file. If this and '--gff3' options are not provided, RSEM will assume 'reference_fasta_files' contains the reference transcripts. In this case, RSEM assumes that name of each sequence in the Multi-FASTA files is its transcript_id.
+        example: annotations.gtf
+      - name: --gff3
+        type: file
+        description: GFF3 annotation file. Converted t GTF format with the file name 'reference_name.gtf'. Please make sure that 'reference_name.gtf' does not exist.
```
Comment (rcannood) [inline src/rsem/rsem_prepare_reference/config.vsh.yaml:48]: fix typo

File: src/rsem/rsem_prepare_reference/config.vsh.yaml [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        description: Assume that 'reference_fasta_files' contains the sequence of a genome, and extract transcript reference sequences using the gene annotations specified in the GTF file. If this and '--gff3' options are not provided, RSEM will assume 'reference_fasta_files' contains the reference transcripts. In this case, RSEM assumes that name of each sequence in the Multi-FASTA files is its transcript_id.
+        example: annotations.gtf
+      - name: --gff3
+        type: file
+        description: GFF3 annotation file. Converted t GTF format with the file name 'reference_name.gtf'. Please make sure that 'reference_name.gtf' does not exist. 
+        example: annotations.gff
+      - name: --gff3_rna_patterns
+        type: string
+        description: Comma-separated list of transcript categories, e.g. "mRNA,rRNA". Only transcripts that match the string will be extracted.
+        example: mRNA,rRNA
```
Comment (rcannood) [inline src/rsem/rsem_prepare_reference/config.vsh.yaml:53]: Should this be made multiple: true?

File: src/rsem/rsem_prepare_reference/config.vsh.yaml [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+      - name: --gff3_genes_as_transcripts
+        type: boolean_true
+        description: This option is designed for untypical organisms, such as viruses, whose GFF3 files only contain genes. RSEM will assume each gene as a unique transcript when it converts the GFF3 file into GTF format.
+      - name: --trusted_sources
+        type: string
+        description: Comma-separated list of trusted sources, e.g. "ENSEMBL,HAVANA". Only transcripts coming from these sources will be extracted. If this option is off, all sources are accepted.
+        example: ENSEMBL,HAVANA
```
Comment (rcannood) [inline src/rsem/rsem_prepare_reference/config.vsh.yaml:60]: Should this be made multiple: true?

File: src/rsem/rsem_prepare_reference/config.vsh.yaml [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        description: Build Bowtie indices. 
+      - name: --bowtie_path 
+        type: string
+        description: The path to the Bowtie executables. (By default the path to Bowtie executables is assumed to be in the user's PATH environment variable)
+      - name: --bowtie2
+        type: boolean_true
+        description: Build Bowtie 2 indices.
+      - name: --bowtie2_path 
+        type: string
+        description: The path to the Bowtie 2 executables. (By default the path to Bowtie 2 executables is assumed to be in the user's PATH environment variable)
```
Comment (rcannood) [inline src/rsem/rsem_prepare_reference/config.vsh.yaml:99]: Remove this -- bowtie is supposed to be installed in the docker image

File: src/rsem/rsem_prepare_reference/config.vsh.yaml [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        description: The length of the poly(A) tails to be added. 
+        example: 125
+      - name: --no_polyA_subset 
+        type: file
+        description: Only meaningful if '--polyA' is specified. Do not add poly(A) tails to those transcripts listed in this file containing a list of transcript_ids.
+        example: transcript_ids.txt
+      - name: --bowtie
+        type: boolean_true
+        description: Build Bowtie indices. 
+      - name: --bowtie_path
```
Comment (rcannood) [inline src/rsem/rsem_prepare_reference/config.vsh.yaml:91]: Remove this -- bowtie is supposed to be installed in the docker image

File: src/rsem/rsem_prepare_reference/config.vsh.yaml [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+      - name: --bowtie2
+        type: boolean_true
+        description: Build Bowtie 2 indices.
+      - name: --bowtie2_path 
+        type: string
+        description: The path to the Bowtie 2 executables. (By default the path to Bowtie 2 executables is assumed to be in the user's PATH environment variable)
+      - name: --star
+        type: boolean_true
+        description: Build STAR indices.
+      - name: --star_path
```
Comment (rcannood) [inline src/rsem/rsem_prepare_reference/config.vsh.yaml:103]: Remove this -- star is supposed to be installed in the docker image

File: src/rsem/rsem_prepare_reference/config.vsh.yaml [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+        type: string
+        description: The path to STAR's executable. (By default the path to Bowtie 2 executables is assumed to be in the user's PATH environment variable)
+      - name: --star_sjdboverhang
+        type: integer
+        description: Length of the genomic sequence around annotated junction. It is only used for STAR to build splice junctions database and not needed for Bowtie or Bowtie2. It will be passed as the --sjdbOverhang option to STAR. According to STAR's manual, its ideal value is max(ReadLength)-1, e.g. for 2x101 paired-end reads, the ideal value is 101-1=100. In most cases, the default value of 100 will work as well as the ideal value. (Default is 100)
+        example: 100
+      - name: --hisat2_hca
+        type: boolean_true
+        description: Build HISAT2 indices on the transcriptome according to Human Cell Atlas (HCA) SMART-Seq2 pipeline.
+      - name: --hisat2_path
```
Comment (rcannood) [inline src/rsem/rsem_prepare_reference/config.vsh.yaml:113]: Remove this -- hisat is supposed to be installed in the docker image

File: src/rsem/rsem_prepare_reference/config.vsh.yaml [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+        description: Build HISAT2 indices on the transcriptome according to Human Cell Atlas (HCA) SMART-Seq2 pipeline.
+      - name: --hisat2_path 
+        type: string
+        description: The path to the HISAT2 executables. (By default the path to HISAT2 executables is assumed to be in the user's PATH environment variable)
+      - name: --quiet
+        alternatives: -q
+        type: boolean_true
+        description: Suppress the output of logging information. 
+  
+  - name: PRIOR-ENHANCED RSEM OPTIONS
```
Comment (rcannood) [inline src/rsem/rsem_prepare_reference/config.vsh.yaml:121]: Fix capitalisation

File: src/rsem/rsem_prepare_reference/test.sh [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+[ ! -d "RSEM_index" ] && echo "RSEM index does not exist!" && exit 1
+[ ! -f "RSEM_index/test.grp" ] && echo "test.grp does not exist!" && exit 1
+[ ! -f "RSEM_index/test.n2g.idx.fa" ] && echo "test.n2g.idx.fa does not exist!" && exit 1
+[ ! -f "RSEM_index/test.ti" ] && echo "test.ti does not exist!" && exit 1
+[ ! -f "RSEM_index/test.idx.fa" ] && echo "test.idx.fa does not exist!" && exit 1
+[ ! -f "RSEM_index/test.seq" ] && echo "test.seq does not exist!" && exit 1
+[ ! -f "RSEM_index/test.transcripts.fa" ] && echo "test.transcripts.fa does not exist!" && exit 1
+
+# echo ">>> Checking whether output is correct"
+# [ ! -f "RSEM_index/test.grp" ] && echo "test.grp does not exist!" && exit 1
```
Comment (rcannood) [inline src/rsem/rsem_prepare_reference/test.sh:29]: Check content of output to see whether the genome.fasta was processed correctly

**Review by rcannood — CHANGES_REQUESTED** [review]

File: src/rsem/rsem_prepare_reference/config.vsh.yaml (hunk 2 above)
Comment (sainirmayi) [reply src/rsem/rsem_prepare_reference/config.vsh.yaml:53]: RSEM also expects a string of comma separated categories so I decided to use the same. In that case it shouldn't be multiple true.

File: src/rsem/rsem_prepare_reference/config.vsh.yaml (hunk 3 above)
Comment (sainirmayi) [reply src/rsem/rsem_prepare_reference/config.vsh.yaml:60]: RSEM also expects a string of comma separated categories so I decided to use the same. In that case it shouldn't be multiple true.

File: src/rsem/rsem_prepare_reference/config.vsh.yaml (hunk 2 above)
Comment (rcannood) [reply src/rsem/rsem_prepare_reference/config.vsh.yaml:57]: I'd still like `multiple` to be set to true. If you replace `;` with `,`, then it should work regardless of whether somebody provides multiple inputs (`--gff3_rna_patterns mRNA --gff3_rna_patterns rRNA`) or one input using the separator of the original tool (`--gff3_rna_patterns mRNA,rRNA`)

File: src/rsem/rsem_prepare_reference/config.vsh.yaml (hunk 3 above)
Comment (rcannood) [reply src/rsem/rsem_prepare_reference/config.vsh.yaml:64]: Same, see above

File: src/rsem/rsem_prepare_reference/config.vsh.yaml [hunk 10]
```diff
@@ -0,0 +1,194 @@
+name: rsem_prepare_reference
+namespace: rsem
+description: | 
+  RSEM is a software package for estimating gene and isoform expression levels from RNA-Seq data. This component prepares transcript references for RSEM.
+keywords: ["Transcriptome", "Index"]
+links:
+  homepage: https://github.com/deweylab/RSEM
+  documentation: https://deweylab.github.io/RSEM/rsem-prepare-reference.html
+  repository: https://github.com/deweylab/RSEM?
```
Comment (rcannood) [inline src/rsem/rsem_prepare_reference/config.vsh.yaml:9]:
```suggestion
  repository: https://github.com/deweylab/RSEM

File: src/rsem/rsem_prepare_reference/config.vsh.yaml [hunk 11]
```diff
@@ -0,0 +1,194 @@
+name: rsem_prepare_reference
+namespace: rsem
+description: | 
+  RSEM is a software package for estimating gene and isoform expression levels from RNA-Seq data. This component prepares transcript references for RSEM.
+keywords: ["Transcriptome", "Index"]
+links:
+  homepage: https://github.com/deweylab/RSEM
```
Comment (rcannood) [inline src/rsem/rsem_prepare_reference/config.vsh.yaml:7]:
```suggestion
  homepage: http://deweylab.github.io/RSEM

File: src/rsem/rsem_prepare_reference/script.sh [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+# replace ';' with ','
+par_reference_fasta_files=$(echo $par_reference_fasta_files | tr ';' ',')
+
+echo "$par_reference_fasta_files"
+rsem-prepare-reference \
+    ${par_gtf:+--gtf "${par_gtf}"} \
+    ${par_gff3:+--gff3 "${par_gff3}"} \
+    ${par_gff3_rna_patterns:+--gff3-RNA-patterns "${par_gff3_rna_patterns}"} \
+    ${par_gff3_genes_as_transcripts:+--gff3-genes-as-transcripts "${par_gff3_genes_as_transcripts}"} \
+    ${par_trusted_sources:+--gtrusted-sourcestf "${par_trusted_sources}"} \
```
Comment (rcannood) [inline src/rsem/rsem_prepare_reference/script.sh:21]: Fix refactoring error

```suggestion
    ${par_trusted_sources:+--trusted-sources "${par_trusted_sources}"} \

**Review by rcannood — CHANGES_REQUESTED** [review]
Minor changes requested

File: src/rsem/rsem_prepare_reference/script.sh [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+    ${par_hisat2_hca:+--hisat2-hca} \
+    ${par_quiet:+--quiet} \
+    ${par_prep_pRSEM:+--prep-pRSEM} \
+    ${par_mappability_bigwig_file:+--mappability-bigwig-file "${par_mappability_bigwig_file}"} \
+    ${meta_cpus:+--num-threads "${meta_cpus}"} \
+    ${par_reference_fasta_files} \
+    $par_reference_name
+
+mkdir -p $par_output
+mv $par_reference_name.* $par_output/
```
Comment (rcannood) [inline src/rsem/rsem_prepare_reference/script.sh:42]:
```suggestion
    "${par_reference_fasta_files}" \
    "${par_reference_name}

mkdir -p "${par_output}"
mv "${par_reference_name}.*" "${par_output}/"

File: src/rsem/rsem_prepare_reference/test.sh [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+cat > genome.fasta <<'EOF'
+>Sheila
+GCTAGCTCAGAAAAaaaNNN
+EOF
+
+echo ">>> Prepare RSEM reference without gene annotations"
+"$meta_executable" \
+  --reference_fasta_files genome.fasta \
+  --reference_name test \
+  --output RSEM_index \
```
Comment (rcannood) [inline src/rsem/rsem_prepare_reference/test.sh:17]:
```suggestion
  --output RSEM_index

File: src/rsem/rsem_prepare_reference/test.sh [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+echo ">>> Checking whether output is correct"
+[ ! -s "RSEM_index/test.grp" ] && echo "test.grp is empty!" && exit 1
+[ ! -s "RSEM_index/test.ti" ] && echo "test.ti is empty!" && exit 1
+[ ! -s "RSEM_index/test.seq" ] && echo "test.seq is empty!" && exit 1
+grep -q "GCTAGCTCAGAAAAaaaNNN" "RSEM_index/test.transcripts.fa" || { echo "THe content of file 'test.transcripts.fa' seems to be incorrect." && exit 1; }
+grep -q "GCTAGCTCAGAAAAAAANNN" "RSEM_index/test.idx.fa" || { echo "FilThe content of file 'test.idx.fa' seems to be incorrect." && exit 1; }
+grep -q "GCTAGCTCAGAAAAAAAGGG" "RSEM_index/test.n2g.idx.fa" || { echo "The contenet of file 'test.n2g.idx.fa' semems to be incorrect." && exit 1; }
+
+echo "All tests succeeded!"
+exit 0
```
Comment (rcannood) [inline src/rsem/rsem_prepare_reference/test.sh:37]:
```suggestion
exit 0

```

File: src/rsem/rsem_prepare_reference/test.sh [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+[ ! -f "RSEM_index/test.seq" ] && echo "test.seq does not exist!" && exit 1
+[ ! -f "RSEM_index/test.transcripts.fa" ] && echo "test.transcripts.fa does not exist!" && exit 1
+
+echo ">>> Checking whether output is correct"
+[ ! -s "RSEM_index/test.grp" ] && echo "test.grp is empty!" && exit 1
+[ ! -s "RSEM_index/test.ti" ] && echo "test.ti is empty!" && exit 1
+[ ! -s "RSEM_index/test.seq" ] && echo "test.seq is empty!" && exit 1
+grep -q "GCTAGCTCAGAAAAaaaNNN" "RSEM_index/test.transcripts.fa" || { echo "THe content of file 'test.transcripts.fa' seems to be incorrect." && exit 1; }
+grep -q "GCTAGCTCAGAAAAAAANNN" "RSEM_index/test.idx.fa" || { echo "FilThe content of file 'test.idx.fa' seems to be incorrect." && exit 1; }
+grep -q "GCTAGCTCAGAAAAAAAGGG" "RSEM_index/test.n2g.idx.fa" || { echo "The contenet of file 'test.n2g.idx.fa' semems to be incorrect." && exit 1; }
```
Comment (rcannood) [inline src/rsem/rsem_prepare_reference/test.sh:34]:
```suggestion
grep -q "GCTAGCTCAGAAAAaaaNNN" "RSEM_index/test.transcripts.fa" || { echo "The content of file 'test.transcripts.fa' seems to be incorrect." && exit 1; }
grep -q "GCTAGCTCAGAAAAAAANNN" "RSEM_index/test.idx.fa" || { echo "The content of file 'test.idx.fa' seems to be incorrect." && exit 1; }
grep -q "GCTAGCTCAGAAAAAAAGGG" "RSEM_index/test.n2g.idx.fa" || { echo "The content of file 'test.n2g.idx.fa' seems to be incorrect." && exit 1; }

**Review by rcannood — APPROVED** [review]
LGTM

**Pull request merged**